ambry-router/src/main/java/com/github/ambry/router/OperationController.java

```diff
@@ -106,13 +107,15 @@ public OperationController(String suffix, String defaultPartitionClass, AccountS
     this.responseHandler = responseHandler;
     this.nonBlockingRouter = nonBlockingRouter;
     // Warm up connections to dataNodes in local and remote DCs.
-    List<ResponseInfo> responseInfos = Collections.synchronizedList(new ArrayList<>());
     String localDatacenter = clusterMap.getDatacenterName(clusterMap.getLocalDatacenterId());
     Map<Boolean, List<DataNodeId>> localAndRemoteNodes = clusterMap.getDataNodeIds()
         .stream()
         // ignore any in-process "data nodes" without TCP ports
         .filter(dataNodeId -> dataNodeId.getPort() != DataNodeId.UNKNOWN_PORT)
         .collect(Collectors.partitioningBy(dataNodeId -> localDatacenter.equals(dataNodeId.getDatacenterName())));
+    // 'responseInfos' list needs to be thread safe since it it possible that the list can be modified as we are reading
+    // it for scenario where we returned from networkClient.warmUpConnections() due to time out.
+    List<ResponseInfo> responseInfos = new CopyOnWriteArrayList<>();
```

Comment:
maybe just use forEach to replace the for loop in line 127. for loop would return a iterator, which is not protected by mutex and requires locking by users. But forEach is protected by the mutex in the SynchronizedList. This way, we won't see the ConcurrentModificationException.

Reply:
+1

Reply:
nvm, i think cow is better in this case iff array is small and writes are few. forEach is limited in the things it can do.

https://stackoverflow.com/questions/16635398/java-8-iterable-foreach-vs-foreach-loop

Reply:
Thank you for the suggestions. For now, used forEach with SynchronizedList() since there is not too much logic in the block and no exceptions to handle etc. 

Since this list is used to keep track of connection failures during warming up connections from frontends to servers, we can probably have more writes if there are lot of server errors and CopyOnWriteArrayList may be sub-optimal. For example, in below, there seem to be around 733 failures/writes during connection warm up.

// local datacenter connection warm up
2022/09/09 19:26:40.290 INFO [Http2NetworkClient] [main] [ambry-frontend] [] HTTP2 connection warm up done. Tried: 2628, Succeeded: 2046, Failed: 549, Time elapsed: 5251 ms
// remote datacenter connection warm up
2022/09/09 19:26:45.475 INFO [Http2NetworkClient] [main] [ambry-frontend] [] HTTP2 connection warm up done. Tried: 930, Succeeded: 743, Failed: 184, Time elapsed: 5185 ms

Reply:
had an offline discussion with arun to discuss pros and cons of foreach vs cow